Updating to k/k v1.20

cici37 opened the pull request:
This PR will:
- Updating cloud-provider-gcp to both deploy with and build against K/K v1.20.
- Updated the libraries we build against.
- Fix the presubmits check and conformance test.

Summary of the changes:
1. Most of the files changes are under /vendor(roughly 1487 files)
2. Removed pkg/apis directory due to credentialprovider moved to staging
3. Removed most code under /cmd/cloud-controller-manager/app/ due to CCM moved to staging
4. Update to k8s.io/klog/v2
5. There are some credential provider related change under /cmd/auth-provider-gcp
6. Changed under /cmd/gcp-controller-manager is mainly due to upstream code changes.
7. Changes under /cluster is for (a)related test script updates (b)/klog/v2 update
8. Update release hash in WORKSPACE
9. Fixed image tagging bug under /tools

cici37 commented:
/assign @cheftako @jpbetz 
Please have a look when you have time. Thank you

/cc @DangerOnTheRanger 
Hi Kermit, I believe here is the place you are going to work on: `https://github.com/kubernetes/cloud-provider-gcp/blob/a0b815787d9b86f9a22ac11acefef66416317aef/cmd/cloud-controller-manager/main.go#L87`

cici37 commented on go.mod:
```diff
@@ -1,79 +1,69 @@
-	k8s.io/cluster-bootstrap v0.19.0 // indirect
-	k8s.io/cri-api v0.19.0 // indirect
+	k8s.io/controller-manager v0.20.0
 	k8s.io/klog v1.0.0
```
Will remove this later.

cheftako commented on cmd/cloud-controller-manager/BUILD:
```diff
@@ -16,14 +16,29 @@ go_binary(
         "//vendor/k8s.io/component-base/metrics/prometheus/version:go_default_library",
+        "//vendor/k8s.io/controller-manager/app:go_default_library",
+        "//vendor/k8s.io/klog/v2:go_default_library",
+        "//vendor/k8s.io/kubernetes/pkg/controller/nodeipam:go_default_library",
```
I get it but yuck. Please make sure we have an issue filed to move these to staging, so we can stop depending on K/K internal.

cici37 replied:
Working on it. :)

cheftako commented on cmd/cloud-controller-manager/main.go:
```diff
@@ -0,0 +1,124 @@
+
+	controllerInitializers := app.DefaultControllerInitializers(c.Complete(), cloud)
+
+	//TODO: passing the right values to nodeIPAMController
```
This was fine in the how to use CCM sample. However here we need the actual node ipam config populated from the command line input parameters. Also please make sure we have correctly wired in the allocator type.

cheftako replied:
(You can leave populating the parameters on the command line to @DangerOnTheRanger but we need to make sure they are wired here)

cheftako commented on cmd/cloud-controller-manager/main.go:
```diff
@@ -0,0 +1,124 @@
+	// the flags could be set before execute
+	command.Flags().VisitAll(func(flag *pflag.Flag) {
+		if flag.Name == "cloud-provider" {
+			flag.Value.Set("SampleCloudProviderFlagValue")
```
?

cheftako commented on cmd/cloud-controller-manager/nodeipamcontroller.go:
```diff
@@ -0,0 +1,210 @@
+
+// setNodeCIDRMaskSizes returns the IPv4 and IPv6 node cidr mask sizes.
+// If --node-cidr-mask-size not set, then it will return default IPv4 and IPv6 cidr mask sizes.
+func setNodeCIDRMaskSizes(cfg nodeipamconfig.NodeIPAMControllerConfiguration) (int, int, error) {
```
I wouldn't want to make NodeIPAMControllerConfiguration part of a utility. However these seem like functions which belong in a utility and not here. We would obviously need to pass in the values from the NodeIPAMControllerConfiguration rather than the whole NodeIPAMControllerConfiguration.

cheftako commented:
/cc @mikedanese to look at the gcp-controller-manager changes.

DangerOnTheRanger commented on cmd/auth-provider-gcp/provider/BUILD:
```diff
@@ -15,9 +15,7 @@ go_library(
         "//pkg/apis/credentialprovider:go_default_library",
```
I think we can remove the entire pkg/apis/credentialprovider folder now that we're supporting 1.20 - this package is now in staging/ at `k8s.io/kubelet/pkg/apis/credentialprovider` as a part of [this PR](https://github.com/kubernetes/kubernetes/pull/94196). This is the only package in the repo that depends upon `pkg/apis/credentialprovider`, I believe, so the entire pkg/apis/credentialprovider directory should be able to be safely removed afterward.

cici37 replied:
Thank you for the reminder. Updated to remove this.

cici37 commented:
/test cloud-provider-gcp-tests

cici37 commented:
/test cloud-provider-gcp-e2e-create

cheftako commented:
One of #181 or #183 is I believe blocked on this.

cici37 commented:
@cheftako I have a temp fix in this PR(Will update after the fix merged to k/k). Could we try to merge this PR first?

cheftako commented on cmd/cloud-controller-manager/main.go:
```diff
@@ -0,0 +1,111 @@
+	if err := command.Execute(); err != nil {
+		os.Exit(1)
+	}
+	//rand.Seed(time.Now().UnixNano())
```
Can we lose this huge chunk of commented out code if we don't need it?

cici37 replied:
Sorry.Updated

cici37 commented:
/test cloud-provider-gcp-e2e-create

cici37 commented:
/test cloud-provider-gcp-e2e-create

cici37 commented:
/test cloud-provider-gcp-e2e-create

cheftako commented:
I believe this specific failure is because your missing https://github.com/kubernetes/kubernetes/pull/94090. Should probably look at making sure you have merged the cluster/gce updates.

cici37 commented:
> I believe this specific failure is because your missing [kubernetes/kubernetes#94090](https://github.com/kubernetes/kubernetes/pull/94090). Should probably look at making sure you have merged the cluster/gce updates.

Thank you for the info. I found the issue is due to config-test did not update with the podpresets change. Updated in latest commit.

cici37 commented:
To make review easier, here are some tips for this huge PR:
1. Most of the files change is under /vendor(roughly 1487)
2. Removed `pkg/apis/credentialprovider` directory due to credentialprovider moved to staging
3. Removed most code under `/cmd/cloud-controller-manager/app/` due to CCM moved to staging
4. Update to `k8s.io/klog/v2`
5. There are some credential provider related change under `/cmd/auth-provider-gcp` 
6. Changed under `/cmd/gcp-controller-manager` is mainly due to upstream code changes.
7. Changed under /cluster is for cluster and related test updates

There is a long stand issue with presubmits failure is due to config-test not updated with `kubernetes/kubernetes#94090`. Fixed in a [separate commit](https://github.com/kubernetes/cloud-provider-gcp/pull/178/commits/268a21c8b3ded1f38b6d7c1b1c99435db91ba67b).

cici37 commented:
Addition info:
1. NodeIPAM controller will be added after upstream fix is in.
2. Conformance tests passed except 6 sig-network failures which is not introduced by this PR. Will work on fixing after NodeIPAM controller is in place.
<img width="1792" alt="Screen Shot 2021-03-10 at 3 16 01 PM" src="https://user-images.githubusercontent.com/8658046/111202999-b93d2680-8581-11eb-89e8-08c39e15551c.png">

jpbetz commented:
Okay, with some help from @cici37 plus the details in https://github.com/kubernetes/cloud-provider-gcp/pull/178#issuecomment-799646763 I was able to break this down.

`go.mod` looks reasonable to me. All kubernetes dependencies have been bumped from v0.19.0 to v0.20.0 as expected.

`git show --numstat 9d44ec61 | grep -v "vendor/"` shows the non-vendor changes:

- `cluster/` changes are purely dependency related and update to use `klog/v2`.
- `cmd/` and `pkg/` credentialprovider move - there's still a `pkg/apis/credentialprovider`directory. Needed?
- `cmd/` CCM move to staging look fine. `cmd/cloud-controller-manager/main.go` and `cmd/cloud-controller-manager/app/controllermanager.go` look good.

- PodsPresets change appears to have been applied to config-helper.sh and config-test in a reasonable way (xref: https://github.com/kubernetes/kubernetes/pull/94090)
- `cluster/bin/kubectl version` reports "v1.20.0" as expected
-  `tools/workspace-status.sh` reports `STABLE_IMAGE_TAG v1.20.0+268a...` as expected

jpbetz commented:
LGTM once we check if `pkg/apis/credentialprovider` can be removed entirely (`pkg/apis/credentialprovider/install/BUILD` is the only remaining file).

jpbetz commented:
/lgtm

cici37 commented:
@cheftako Gentle reminder on the review😄

cheftako commented on cluster/addons/fluentd-elasticsearch/es-image/BUILD:
```diff
@@ -22,7 +22,7 @@ go_library(
         "//vendor/k8s.io/client-go/rest:go_default_library",
         "//vendor/k8s.io/client-go/tools/clientcmd:go_default_library",
         "//vendor/k8s.io/client-go/tools/clientcmd/api:go_default_library",
-        "//vendor/k8s.io/klog:go_default_library",
```
Flag for @liggitt. Want to make sure he's aware we're switching to klog/v2 as it has caused issues in the past.

cheftako commented on cmd/cloud-controller-manager/app/controllermanager.go:
```diff
@@ -1,59 +1,26 @@
-/*
-Copyright 2016 The Kubernetes Authors.
```
We still need a K8s Author copyright notice.

cici37 replied:
Added back. Thanks

jpbetz replied:
We need to lint for this in this repo. I'll track that separately.

cheftako commented on go.mod:
```diff
@@ -1,66 +1,72 @@
+	k8s.io/api v0.20.0
+	k8s.io/apimachinery v0.20.0
+	k8s.io/apiserver v0.20.0
+	k8s.io/client-go v9.0.0+incompatible
```
cc @caesarxuchao but this client-go version seems wrong.

jpbetz replied:
v0.20.0 looks congruent? https://github.com/kubernetes/client-go/releases/tag/v0.20.0

cici37 replied:
I assume Line 12 `k8s.io/client-go => k8s.io/client-go v0.20.0` will get the right version for client go?

jpbetz replied:
Yes, the `replace` will result in the correct version. It would be nice to have all the `require` statements aligned on `v0.20.0`, just to minimize confusion. Does something like `go get k8s.io/client@v0.20.0` will update that line appropriately?

cici37 replied:
Updated go.mod in the latest commit.

cheftako commented on go.mod:
```diff
@@ -1,66 +1,72 @@
+	k8s.io/cloud-provider v0.20.0
+	k8s.io/component-base v0.20.0
+	k8s.io/kubelet v0.20.0
+	k8s.io/kubernetes v1.20.0
```
It would be good if we could remove the k8s.io/kubernetes dependency. Do we know why we have it? We should not be relying on anything K/K private.

cici37 replied:
Updated. (It was there because previous NodeIPAM related dependencies. Sorry)

jpbetz commented:
/lgtm

cheftako commented on cmd/cloud-controller-manager/app/controllermanager.go:
```diff
@@ -68,13 +51,15 @@ the cloud specific control loops shipped with Kubernetes.`,
 			cliflag.PrintFlags(cmd.Flags())
 
-			c, err := s.Config(KnownControllers(), ControllersDisabledByDefault.List())
+			c, err := s.Config(knownControllers(), app.ControllersDisabledByDefault.List())
```
app.ControllersDisabledByDefault.List() is probably not what we want to do. This is deferring responsibility for the set of controllers to be deferred to K/K rather then gcp. A concrete example is we may want to disable on of the controllers when we turn on import NodeIPAMController. If we ever end up switching to our own ServiceController, then we will want to disable the default one.  So this is fine for now but we should plan to switch to a list we control here.

cici37 replied:
Those will be addressed in v1.21.  Thanks

cheftako commented on cmd/cloud-controller-manager/app/controllermanager.go:
```diff
@@ -83,7 +68,7 @@ the cloud specific control loops shipped with Kubernetes.`,
 
 	fs := cmd.Flags()
-	namedFlagSets := s.Flags(KnownControllers(), ControllersDisabledByDefault.List())
+	namedFlagSets := s.Flags(knownControllers(), app.ControllersDisabledByDefault.List())
```
Ditto.

cheftako commented on go.mod:
```diff
@@ -1,66 +1,71 @@
+	sigs.k8s.io/kubetest2 v0.0.0-20210309183806-9230b4e73d8d // indirect
 )
 
 go 1.13
```
Not for now but we should try to keep up with K/K. That suggests this needs to be upgraded. Should coordinate that change with @mikedanese, especially given the https://golang.org/doc/go1.15#commonname change.

cici37 replied:
The issue to track: https://github.com/kubernetes/cloud-provider-gcp/issues/194

cheftako commented on tools/version.sh:
```diff
@@ -74,7 +74,7 @@ get_version_vars() {
     # Use git describe to find the version based on tags.
     #if [[ -n ${KUBE_GIT_VERSION-} ]] || KUBE_GIT_VERSION=$("${git[@]}" describe --tags --match='v*' --abbrev=14 "${KUBE_GIT_COMMIT}^{commit}" 2>/dev/null); then
-    if [[ -n ${KUBE_GIT_VERSION-} ]] || KUBE_GIT_VERSION=$(echo "v1.18.0-alpha+${KUBE_GIT_COMMIT}"); then
+    if [[ -n ${KUBE_GIT_VERSION-} ]] || KUBE_GIT_VERSION=$(echo "v1.20.0+${KUBE_GIT_COMMIT}"); then
```
Is there a reason we removed the "-alpha" suffix? I've generally found it useful to indicate a local build.

cheftako commented:
/lgtm
/approve
Couple of comments of things to look for as we move forward but nothing to hold up what you have.
